rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,430 @@
+  is aware of all memory accesses occurring inside said memory block.
+- [Cryptography](https://en.wikipedia.org/wiki/Cryptography), where it is
+  extremely important to ensure that sensitive information is erased after use,
+  and is not leaked via indirect means such as recognizable scaling patterns in
+  the time taken by a system to process attacker-crafted inputs.
+
+In such circumstances, it may be necessary to assert precise manual control on
+the memory accesses that are carried out by a Rust program in a certain memory
+region. This is the purpose of _volatile memory operations_, which allow a Rust
+programmers to generate a carefully controlled stream of hardware memory load
+and store instructions, which is guaranteed to be left untouched by the Rust
+compiler's optimizer even though surrounding Rust code will continue to be
+optimized as usual.
+
+---
+
+Volatile memory operations are exposed in the `std::volatile` module of the Rust
+standard library, or alternatively in the `core::volatile` module for the
+purpose of writing `#![no_std]` applications. They are interfaced through
+fixed-size data wrappers that are somewhat reminiscent of the API used for
+atomic operations in `std::sync::atomic`:
+
+```rust
+use std::sync::atomic::Ordering;
+use std::ptr::NonNull;
+use std::volatile::VolatileU8;
+
+unsafe fn do_volatile_things(target: NonNull<VolatileU8>) -> u8 {
+    target.store(42, Ordering::Relaxed);
+    target.load_not_atomic()
+}
+```
+
+Several specificities, however, should be apparent from the above usage example.
+
+First of all, volatile types must be manipulated via pointers, instad of Rust
+references. These unusual and unpleasant ergonomics are necessary in order to
+achieve the desired semantics of manually controlling every access to the target
+memory location, because the mere existence of a Rust reference pointing to a
+memory region allows the Rust compiler to generate memory operations targeting
+this region (be they prefetches, register spills...).
+
+Because a Rust pointer is not subjected to borrow checking and has no obligation
+of pointing towards a valid memory location, this means that using a Volatile
+wrapper in any way is unsafe.
+
+Second, in addition to familiar atomic memory operations, volatile types expose
+the `load_not_atomic()` and `store_not_atomic()` methods. As their name suggest,
+these memory operations are not considered to be atomic by the compiler, and
+are therefore not safe to concurrently invoke in multiple threads.
+
+On the vast majority of hardware platforms supported by Rust, using these
+methods will generate exactly the same code as using the `load()` and `store()`
+methods with `Relaxed` memory ordering, with the only difference being that
+data races are Undefined Behavior. When that is the case, safer `Relaxed` atomic
+volatile operations should be preferred to their non-atomic equivalents.
+
+But unfortunately, Rust supports a couple of platforms, such as the `nvptx`
```

Comment:
I do not have such a list, and I am not sure who I should ping to get it. Previous discussion mentioned two broad categories of targets where this might happen:

- Hardware without global cache coherence, where local caches are allowed to get out of sync with DRAM (this is `nvptx`, and probably other GPUs as well).
- Emulated or abstract hardware, including language virtual machines like WASM and the JVM.

Reply:
All targets that rust supports are in https://github.com/rust-lang/rust/tree/master/src/librustc_target/spec, so that would be a nice place to inspect. It might well be that these are the only such targets.

Reply:
I took a look through that list and did some quick research and godbolt experimentations, but did not find any architecture in this list which doesn't either 1/have native loads and stores with `Relaxed` or stronger semantics or 2/belongs to one of the two categories above. So I think that the above characterization is accurate.

More precisely...

- I think that arm/thumb, hexagon, mips, msp340, power, riscv, s390x, sparc and x86 either provide sufficiently strong cache coherence between cores for `Relaxed` ordering to be free or only have single-core implementations.
- I think that `nvptx64` is the only real hardware architecture in that list with weaker-than-`Relaxed` default cache coherence, by virtue of not being globally cache-coherent. But I would expect AMD GPUs to be similar if we ever add support for them.
- `asmjs` and `wasm32` are the only examples of abstract hardware in that list, but I would expect the JVM to behave similarly if we ever end up targeting that.